### PR TITLE
feat: add empty states to sqlab editor and select

### PR DESCRIPTION
**Pull request by pkdotson**
### SUMMARY
This pr creates an empty state for the dropdown for the table selector and query editor when there is no database selected.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
after

https://user-images.githubusercontent.com/17326228/162277445-cefc2bdf-6ee5-4ce4-912d-dafb85dda4fc.mov

### TESTING INSTRUCTIONS

Go to sql editor. If there is no db selected to query there should be an empty state for the query editor. If there is no db connected to db select dropdown should show empty state with link to foward user to db listview.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
@pkdotson The Select has a property called `notFoundContent` that can be used to implement this behavior. Keep in mind that if a user searches for a database and none is found, the same screen will appear. So I recommend a more generic message like "No databases found" instead of "There are no database available"

**Comment by michael-s-molina** [comment]
@pkdotson Tip: If you leave a blank line before the video URL in the PR description, Github will show the video instead of the URL.

```
after
<blank_line> <--- insert a blank line
https://user-images.githubusercontent.com/17326228/162277445-cefc2bdf-6ee5-4ce4-912d-dafb85dda4fc.mov
```

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
@pkdotson The `DatabaseSelector` is a generic component meant to be used by multiple modules. That's why it's inside `src/components`. You shouldn't add SQL Lab specific logic inside this component like you're doing when interacting with Redux. All this logic should be encapsulated in a SQL Lab component that uses the `DatabaseSelector` and pass a `notFoundContent` property when applicable.

**Comment by pkdotson** [comment]
> @pkdotson The `DatabaseSelector` is a generic component meant to be used by multiple modules. That's why it's inside `src/components`. You shouldn't add SQL Lab specific logic inside this component like you're doing when interacting with Redux. All this logic should be encapsulated in a SQL Lab component that uses the `DatabaseSelector` and pass a `notFoundContent` property when applicable.

But this component is only being used in Sqllab and used in changing the database. It makes sense for the action to happen here as this where that action is happening.

**Comment by michael-s-molina** [comment]
> But this component is only being used in Sqllab and used in changing the database. It makes sense for the action to happen here as this where that action is happening.

This component is also used by Explore and the dataset listing.

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx [hunk 1]
```diff
@@ -92,6 +93,7 @@ export default function SqlEditorLeftBar({
   // Ref needed to avoid infinite rerenders on handlers
   // that require and modify the queryEditor
   const queryEditorRef = useRef<QueryEditor>(queryEditor);
+  const [noDbFoundTitle, setTitleForNoDBFound] = useState(false);
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:96]: can we change this to `isDBConnected` instead?

Comment (betodealmeida) [reply superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:96]: Yeah, the name `noDbFoundTitle` makes me thing this is a string, not a boolean.

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx [hunk 2]
```diff
@@ -142,6 +144,22 @@ export default function SqlEditorLeftBar({
...
+    <EmptyStateSmall
+      image="empty.svg"
+      title={
+        noDbFoundTitle
+          ? 'No databases match your search'
+          : 'There are no databases available'
+      }
+      description={
+        <p>
+          Manage your database <a href="/databaseview/list">here</a>
```
Comment (betodealmeida) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:157]: These should be translatable with `t()`.

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx (hunk 1 above)
Comment (pkdotson) [reply superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:96]: Yep

File: superset-frontend/src/SqlLab/components/SqlEditor/index.jsx [hunk 3]
```diff
@@ -227,7 +231,11 @@ class SqlEditor extends React.PureComponent {
     // We need to measure the height of the sql editor post render to figure the height of
     // the south pane so it gets rendered properly
     // eslint-disable-next-line react/no-did-mount-set-state
+    const db = this.props.database;
     this.setState({ height: this.getSqlEditorHeight() });
+    if (!db || isEmpty(db)) {
+      this.setState({ showEmptyState: true });
```
Comment (michael-s-molina) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:237]:
```suggestion
      this.setEmptyState(true);
```

File: superset-frontend/src/SqlLab/components/SqlEditor/index.jsx [hunk 4]
```diff
@@ -753,10 +765,19 @@ class SqlEditor extends React.PureComponent {
...
+              setEmptyState={this.setEmptyState}
             />
           </div>
         </CSSTransition>
-        {this.queryPane()}
+        {this.state.showEmptyState ? (
+          <EmptyStateBig
+            image="vector.svg"
+            title="Select a database to write a query"
+            description="Choose one of the available databases from the panel on the left."
```
Comment (michael-s-molina) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:776]:
```suggestion
            title={t("Select a database to write a query")}
            description={t("Choose one of the available databases from the panel on the left.")}
```

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx [hunk 5]
```diff
@@ -88,15 +99,29 @@ export default function SqlEditorLeftBar({
   queryEditor,
   tables = [],
   height = 500,
+  setEmptyState,
 }: SqlEditorLeftBarProps) {
   // Ref needed to avoid infinite rerenders on handlers
   // that require and modify the queryEditor
   const queryEditorRef = useRef<QueryEditor>(queryEditor);
+  const [isDbSearch, setDbSearch] = useState(false);
+  const [isSelectIsClosed, setSelectIsClosed] = useState(true);
```
Comment (michael-s-molina) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:108]:
```suggestion
```

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx [hunk 6]
```diff
@@ -88,15 +99,29 @@ export default function SqlEditorLeftBar({
...
+
   useEffect(() => {
     queryEditorRef.current = queryEditor;
   }, [queryEditor]);
 
+  useEffect(() => {
+    if (isSelectIsClosed) setDbSearch(false);
+  }, [isSelectIsClosed]);
```
Comment (michael-s-molina) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:116]:
```suggestion
```

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx [hunk 7]
```diff
@@ -185,6 +227,10 @@ export default function SqlEditorLeftBar({
   return (
     <div className="SqlEditorLeftBar">
       <TableSelectorMultiple
+        onEmptyResults={onDbSearch}
+        emptyState={emptyStateComponent}
+        setEmptyState={setEmptyState}
+        dbSelectClosedState={setSelectIsClosed}
```
Comment (michael-s-molina) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/index.tsx:233]:
```suggestion
```

File: superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx [hunk 8]
```diff
@@ -224,6 +224,18 @@ test('Should database select display options', async () => {
...
+test('should show empty state if there are no options', async () => {
+  const props = createProps();
+  // @ts-ignore
+  props.db = null;
+  render(<DatabaseSelector {...props} />, { useRedux: true });
+  const select = screen.getByRole('combobox', {
+    name: 'Select database or type database name',
+  });
+  userEvent.click(select);
+  expect(screen.queryByText('test-mysql')).not.toBeInTheDocument();
```
Comment (michael-s-molina) [inline superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx:236]: Right now we are only checking for the absence of the test-mysql text but there is no check about the empty text content.

File: superset-frontend/src/components/DatabaseSelector/index.tsx [hunk 9]
```diff
@@ -88,14 +95,17 @@ type SchemaValue = { label: string; value: string };
...
   formMode?: boolean;
   getDbList?: (arg0: any) => {};
   handleError: (msg: string) => void;
   isDatabaseSelectEnabled?: boolean;
   onDbChange?: (db: DatabaseObject) => void;
+  onEmptyResults?: (searchText?: string) => null;
   onSchemaChange?: (schema?: string) => void;
   onSchemasLoad?: (schemas: Array<object>) => void;
   readOnly?: boolean;
+  setDbSearch?: Dispatch<SetStateAction<boolean>>;
```
Comment (michael-s-molina) [inline superset-frontend/src/components/DatabaseSelector/index.tsx:108]:
```suggestion
```

File: superset-frontend/src/components/TableSelector/index.tsx [hunk 10]
```diff
@@ -92,6 +95,8 @@ interface TableSelectorProps {
   onTablesLoad?: (options: Array<any>) => void;
   readOnly?: boolean;
   schema?: string;
+  setEmptyState?: Dispatch<SetStateAction<boolean>>;
```
Comment (michael-s-molina) [inline superset-frontend/src/components/TableSelector/index.tsx:98]:
```suggestion
```

File: superset-frontend/src/components/TableSelector/index.tsx [hunk 11]
```diff
@@ -22,6 +22,8 @@ import React, {
   ReactNode,
   useMemo,
   useEffect,
+  Dispatch,
+  SetStateAction,
```
Comment (michael-s-molina) [inline superset-frontend/src/components/TableSelector/index.tsx:26]:
```suggestion
```

File: superset-frontend/src/components/DatabaseSelector/index.tsx [hunk 12]
```diff
@@ -88,11 +88,13 @@ type SchemaValue = { label: string; value: string };
 
 interface DatabaseSelectorProps {
```
Comment (michael-s-molina) [inline superset-frontend/src/components/DatabaseSelector/index.tsx:89]:
```suggestion
export interface DatabaseSelectorProps {
```

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thanks for addressing the comments @pkdotson! I added more non-blocking suggestions regarding the names of the variables.

Here's the Git diff to fix the `DatabaseSelector.test.tsx`

```
diff --git a/superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx b/superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx
index de152cacb..272249b54 100644
--- a/superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx
+++ b/superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx
@@ -21,12 +21,12 @@ import React from 'react';
 import { render, screen, waitFor } from 'spec/helpers/testing-library';
 import { SupersetClient } from '@superset-ui/core';
 import userEvent from '@testing-library/user-event';
-import DatabaseSelector from '.';
+import DatabaseSelector, { DatabaseSelectorProps } from '.';
 import { EmptyStateSmall } from '../EmptyState';
 
 const SupersetClientGet = jest.spyOn(SupersetClient, 'get');
 
-const createProps = () => ({
+const createProps = (): DatabaseSelectorProps => ({
   db: {
     id: 1,
     database_name: 'test',
@@ -39,12 +39,10 @@ const createProps = () => ({
   schema: undefined,
   sqlLabMode: true,
   getDbList: jest.fn(),
-  getTableList: jest.fn(),
   handleError: jest.fn(),
   onDbChange: jest.fn(),
   onSchemaChange: jest.fn(),
   onSchemasLoad: jest.fn(),
-  onUpdate: jest.fn(),
 });
 
 beforeEach(() => {
@@ -192,12 +190,10 @@ test('Refresh should work', async () => {
   await waitFor(() => {
     expect(SupersetClientGet).toBeCalledTimes(2);
     expect(props.getDbList).toBeCalledTimes(0);
-    expect(props.getTableList).toBeCalledTimes(0);
     expect(props.handleError).toBeCalledTimes(0);
     expect(props.onDbChange).toBeCalledTimes(0);
     expect(props.onSchemaChange).toBeCalledTimes(0);
     expect(props.onSchemasLoad).toBeCalledTimes(0);
-    expect(props.onUpdate).toBeCalledTimes(0);
   });
 
   userEvent.click(screen.getByRole('button', { name: 'refresh' }));
@@ -205,12 +201,10 @@ test('Refresh should work', async () => {
   await waitFor(() => {
     expect(SupersetClientGet).toBeCalledTimes(3);
     expect(props.getDbList).toBeCalledTimes(1);
-    expect(props.getTableList).toBeCalledTimes(0);
     expect(props.handleError).toBeCalledTimes(0);
     expect(props.onDbChange).toBeCalledTimes(0);
     expect(props.onSchemaChange).toBeCalledTimes(0);
     expect(props.onSchemasLoad).toBeCalledTimes(2);
-    expect(props.onUpdate).toBeCalledTimes(0);
   });
 });
 
@@ -226,17 +220,15 @@ test('Should database select display options', async () => {
 });
 
 test('should show empty state if there are no options', async () => {
+  SupersetClientGet.mockImplementation(
+    async () => ({ json: { result: [] } } as any),
+  );
   const props = createProps();
-  // @ts-ignore
-  props.db = null;
-  // @ts-ignore
-  props.results = [];
-  const { getByAltText } = await render(
+  render(
     <DatabaseSelector
       {...props}
-      emptyState={
-        <EmptyStateSmall title="testing" description="testing" image="" />
-      }
+      db={undefined}
+      emptyState={<EmptyStateSmall title="empty" image="" />}
     />,
     { useRedux: true },
   );
@@ -244,7 +236,7 @@ test('should show empty state if there are no options', async () => {
     name: 'Select database or type database name',
   });
   userEvent.click(select);
-  const emptystate = getByAltText('empty');
+  const emptystate = await screen.findByText('empty');
   expect(emptystate).toBeInTheDocument();
   expect(screen.queryByText('test-mysql')).not.toBeInTheDocument();
 });
```

File: superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx [hunk 13]
```diff
@@ -224,6 +219,29 @@ test('Should database select display options', async () => {
...
+      db={undefined}
+      emptyState={<EmptyStateSmall title="empty" image="" />}
+    />,
+    { useRedux: true },
+  );
+  const select = screen.getByRole('combobox', {
+    name: 'Select database or type database name',
+  });
+  userEvent.click(select);
+  // const emptystate = getByAltText('empty');
```
Comment (michael-s-molina) [inline superset-frontend/src/components/DatabaseSelector/DatabaseSelector.test.tsx:239]:
```suggestion
```

**Comment by sadpandajoe** [comment]
🏷️  preset:2022.15

**Pull request merged**